File: bms/go/go-gt/cost/cost.go
package test

var ()

const ()

func cost(a *Matrix, b *Matrix, p Vector) (c int64) {
	var i, j int64
	c = 0
	for i = 0; i < p.Len(); i++ {
		for j = 0; j < p.Len(); j++ {
			c += a.Get(i, j) * b.Get(p[i], p[j])
		}
	}
	return c
}

func (v Vector) Len() int64 {
	return int64(len(v))
}

func (m Matrix) Get(i int64, j int64) int64 {
	return m.A[i*m.N+j]
}

type Vector []int64

type Matrix struct {
	N int64
	A []int64
}

File: translations_only/success/go/go-gt/cost/mistral/Hinted/cost.rs


use std::boxed::Box;

type Vector = Vec<i64>;
type Matrix = Box<[(i64, i64)]>;

fn cost(a: &Matrix, b: &Matrix, p: &Vector) -> i64 {
    let mut c = 0;
    let len = p.len() as i64;
    for i in 0..len {
        for j in 0..len {
            let index_a = (i * len + j) as usize;
            let p_i = p[i as usize] as usize;
            let p_j = p[j as usize] as usize;
            let index_b = p_i * len as usize + p_j;
            c += a[index_a].0 * b[index_b].0;
        }
    }
    c
}

fn len(v: &Vector) -> usize {
    v.len()
}

fn get(m: &Matrix, i: i64, j: i64) -> i64 {
    m[(i * m.len() as i64 + j) as usize].0
}

